[Scout] Test config manifests

dolaru opened the pull request:
Because Scout is based on Playwright, detailed information about tests is only available while Playwright is running. This means that any time we need test-level information from a test config file we have to start a Playwright process which can be quite slow and resource-intensive if you want to do this for a larger number of test configs.

These changes are trying to address that by introducing a system where this runtime-only information is stored in manifest files, in close proximity of test config files, automatically kept up to date by CI.

### Usage
Locally, run the new kbn-scout CLI command (via `node scripts/scout update-test-config-manifests`) to regenerate manifests.
  - Running without `--includingUpToDate` will only regenerate manifests that appear outdated (component directory SHA mismatch)
  - Use `--includingUpToDate` to force regenerate all manifests

In Buildkite, `node scripts/scout update-test-config-manifests` runs as a quick check and any changes will be auto-commited.

### Notable changes
- Add Scout reporting package with a registry of test configs and components
- Add a Playwright reporter that writes manifest files that capture test information for static analysis
- Add a Scout CLI command `update-test-config-manifests` to create/update manifests
- Add a versatile Playwright CLI wrapper for Scout tooling to call Playwright programmatically
- Add Buildkite quick check to run manifest updates and commit any changes as `[Scout] Automated config manifest updates`
- Add new path constants/globs/regex for discovering Scout configs
- Changed how test IDs are calculated for Scout: combine test file path and test title when hashing
- Remove project name and test file path from Scout test titles 



<!--ONMERGE {"backportTargets":["8.19","9.1","9.2"]} ONMERGE-->

dolaru committed:
[Scout] Test config manifests

Because Scout is based on Playwright, detailed information about tests is only available while Playwright is running.
This means that any time we need test-level information from a test config file we have to start a Playwright process
which can be quite slow and resource-intensive if you want to do this for a larger number of tests configurations.

These changes are trying to address that by introducing a system where this runtime-only information is stored in
manifest files, stored in the proximity of test config files, automatically kept up to date by CI.

### Usage
Locally, run the new kbn-scout CLI command (via `node scripts/scout update-test-config-manifests`) to regenerate manifests.
  - Running without `--includingUpToDate` will only regenerate manifests that appear outdated (component directory SHA mismatch)
  - Use `--includingUpToDate` to force regenerate all manifests

In Buildkite, `node scripts/scout update-test-config-manifests` runs as a quick check and any changes will be auto-commited.

### Notable changes
- Add Scout reporting package with a registry of test configs and components
- Add a Playwright reporter that writes manifest files that capture test information for static analysis
- Add a Scout CLI command `update-test-config-manifests` to create/update manifests
- Add a versatile Playwright CLI wrapper for Scout tooling to call Playwright programmatically
- Add Buildkite quick check to run manifest updates and commit any changes as `[Scout] Automated config manifest updates`
- Add new path constants/globs/regex for discovering Scout configs
- Changed how test IDs are calculated for Scout: combine test file path and test title when hashing
- Remove project name and test file path from Scout test titles

kibanamachine committed:
[Scout] Automated config manifest updates

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/368665)
* Commit: f796004e3488bd75b41efb48f55349e1c0f3d22d

### Failed CI Steps
* [Build Kibana Distribution](https://buildkite.com/elastic/kibana-pull-request/builds/368665#019ae59c-9185-48ba-9018-2a67df0448c1)

### History
* :broken_heart: [Build #368631](https://buildkite.com/elastic/kibana-pull-request/builds/368631) failed 6bd669709cd41c7e6ea08636b4efdfa354528078

cc @dolaru

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368665","number":368665,"commit":"f796004e3488bd75b41efb48f55349e1c0f3d22d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368631","number":368631,"commit":"6bd669709cd41c7e6ea08636b4efdfa354528078"}],"number":368665}
buildkite-pr-comment-->

dolaru committed:
Merge branch 'main' into scout-mfs

dolaru committed:
fix: test props; title and test end results

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/368935)
* Commit: c9db057718b9cfa73a640b78d081cd1b2fdd44ec

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/368935#019ae913-cad2-456f-9ec4-3f8f3cc00564)

### History
* :yellow_heart: [Build #368665](https://buildkite.com/elastic/kibana-pull-request/builds/368665) was flaky f796004e3488bd75b41efb48f55349e1c0f3d22d
* :broken_heart: [Build #368631](https://buildkite.com/elastic/kibana-pull-request/builds/368631) failed 6bd669709cd41c7e6ea08636b4efdfa354528078

cc @dolaru

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368935","number":368935,"commit":"c9db057718b9cfa73a640b78d081cd1b2fdd44ec"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368665","number":368665,"commit":"f796004e3488bd75b41efb48f55349e1c0f3d22d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368631","number":368631,"commit":"6bd669709cd41c7e6ea08636b4efdfa354528078"}],"number":368935}
buildkite-pr-comment-->

kibanamachine committed:
[Scout] Automated config manifest updates

dmlemeshko commented on src/platform/packages/private/kbn-scout-reporting/src/reporting/playwright/failed_test/failed_test_reporter.ts:
```diff
@@ -115,10 +115,11 @@ export class ScoutFailedTestReporter implements Reporter {
       return;
     }
 
+    const testTitle = test.titlePath().slice(3).join(' ');
```
please add a comment to explain split

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-entity-store-performance-from-pr --> Click to trigger **kibana-entity-store-performance-from-pr** for this PR!

dmlemeshko commented on src/platform/packages/private/kbn-scout-reporting/src/registry/test_config.ts:
```diff
@@ -0,0 +1,140 @@
+      ? {
+          path: manifestPath,
+          exists: manifestExists,
+          ...JSON.parse(readFileSync(manifestPath, 'utf8')),
```
_question_: can me move it above and validate if it is a JSON? just in case we got "corrupted", would good to see a nice error.

dmlemeshko commented on src/platform/packages/private/kbn-scout-reporting/src/registry/test_config.ts:
```diff
@@ -0,0 +1,140 @@
+          path: manifestPath,
+          exists: manifestExists,
+          tests: [],
+        };
```
in `ScoutConfigManifest` 2 field are non-optional, but not defined here:

```
  lastModified: string;
  sha1: string;
```

Do we want to define defaults or change interface?

dolaru replied:
Added defaults:
```js
{
  lastModified: new Date(0).toISOString(),
  sha1: '000000000000000-000000000000000',
}
```

dmlemeshko commented on src/platform/packages/private/kbn-scout-reporting/src/registry/test_config.ts:
```diff
@@ -0,0 +1,140 @@
+    return type === undefined
+      ? configs
+      : configs.filter((config) => config.component.type === type);
+  },
```
```suggestion
  forComponent(name: string, type?: ScoutTestableComponent['type']): ScoutTestConfig[] {
    const filtered = this.all.filter(config => config.component.name === name);
    return type ? filtered.filter(config => config.component.type === type) : filtered;
  },
```

dmlemeshko committed:
Merge branch 'main' into scout-mfs

pheyos commented on src/platform/packages/private/kbn-scout-reporting/src/reporting/playwright/events/playwright_reporter.ts:
```diff
@@ -66,6 +72,7 @@ export class ScoutPlaywrightReporter implements Reporter {
 
     this.name = this.reporterOptions.name || 'unknown';
     this.runId = this.reporterOptions.runId || generateTestRunId();
+    this.captureSteps = this.reporterOptions.captureSteps || false;
```
I think it's fine for now. We can enable it once we actually need it.

dolaru committed:
Tests for test ID generation helpers

dolaru committed:
Merge branch 'refs/heads/main' into scout-mfs

dolaru committed:
Merge branch 'refs/heads/main' into scout-mfs

dolaru committed:
fix: correct title in failed test reporter

dolaru committed:
PR feedback changes

- add tests for `test_config` module
- add missing `lastModified` and `sha1` props to empty manifest object
- readability improvements
- rename, component -> module

dolaru committed:
15 chars per dummy sha part, not 16

dolaru committed:
shorter declaration

dolaru committed:
shorter var names

dolaru committed:
better logging

dolaru committed:
Merge branch 'main' into scout-mfs

dolaru committed:
Manifest updates

dmlemeshko committed:
Merge branch 'main' into scout-mfs

dolaru committed:
fix: make testConfig.fromPath accept both absolute & relative paths

dolaru committed:
Merge branch 'main' into scout-mfs

kibanamachine committed:
[Scout] Automated config manifest updates

dmlemeshko reviewed (approved):
LGTM

tylersmalley commented on src/platform/packages/shared/kbn-scout/test/scout/.meta/ui/standard.json:
```diff
@@ -0,0 +1,132 @@
+{
+  "lastModified": "2025-12-03T19:04:20.842Z",
+  "sha1": "cf987cc228a08cc2083d5d705322a88092e4ba5a",
```
Do we need the lastModified and sha1 properties here? Want to make sure it's something we need as it will likely be a cause of merge conflicts.

dolaru replied:
I guess I could get rid of `lastModified`,  but `sha1` is definitely something that must stay. It's needed to determine if the manifest is still accurately representing the scope of the test config it's describing.

rmyz reviewed (approved):
presentation changes LGTM

dmlemeshko committed:
Merge branch 'main' into scout-mfs

kibanamachine committed:
[Scout] Automated config manifest updates

dolaru commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|9.2|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/245976">](https://github.com/elastic/kibana/pull/245976)|
|✅|9.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/245978">](https://github.com/elastic/kibana/pull/245978)|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/245981">](https://github.com/elastic/kibana/pull/245981)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 10.2.0 -->

kibanamachine commented:
Looks like this PR has backport PRs but they still haven't been merged. Please merge them ASAP to keep the branches relatively in sync.
cc: @dolaru
